File: backend/src/csp_solver/solver/local_search.py
"""Local search (min-conflicts) for CSP solving."""

from __future__ import annotations

import random
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from csp_solver.solver.csp import CSP

type Solution = dict[Any, Any]


def num_conflicts(csp: CSP, v: Any, d: Any, solution: Solution) -> int:
    """Count constraint violations for assigning value d to variable v."""
    count = 0
    old_val = solution.get(v)
    solution[v] = d
    for constraint in csp.constraints[v]:
        if not constraint(solution):
            count += 1
    if old_val is not None:
        solution[v] = old_val
    elif v in solution:
        del solution[v]
    return count


def conflicting_variables(csp: CSP, solution: Solution) -> list[Any]:
    """Return all variables with at least one constraint violation."""
    return [
        v
        for v in csp.variables
        if v in solution and num_conflicts(csp, v, solution[v], solution) > 0
    ]


def min_conflicting_value(csp: CSP, v: Any, solution: Solution) -> Any:
    """Find domain value with minimum conflicts using pure Python min()."""
    return min(csp.domains[v], key=lambda d: num_conflicts(csp, v, d, solution))


def min_conflicts(csp: CSP, iteration_count: int = 10000) -> bool:
    """Hill-climbing local search: randomly initialize, iteratively fix conflicts."""
    solution: Solution = {}
    random.shuffle(csp.variables)

    for v in csp.variables:
        solution[v] = min_conflicting_value(csp, v, solution)

    for _ in range(iteration_count):
        conflicted = conflicting_variables(csp, solution)

        if len(conflicted) == 0:
            csp.solutions.append(solution.copy())
            return True

        v = random.choice(conflicted)
        d = min_conflicting_value(csp, v, solution)
        solution[v] = d

    return False

File: csp-solver/src/solver/local_search.rs
//! Min-conflicts hill-climbing local search.

use crate::adjacency::Adjacency;
use crate::constraint::{ConstraintEnum, VarId};
use crate::domain::Domain;
use crate::variable::Variable;

// ---------------------------------------------------------------------------
// Simple LCG PRNG (no external deps)
// ---------------------------------------------------------------------------

struct Rng(u64);

impl Rng {
    fn new() -> Self {
        let seed = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos() as u64;
        Self(seed)
    }

    fn next(&mut self) -> u64 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        self.0
    }

    fn usize(&mut self, bound: usize) -> usize {
        (self.next() % bound as u64) as usize
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Count the number of constraints violated if `var` is assigned `val` given
/// the current `assignment`.
fn num_conflicts<D: Domain>(
    var: VarId,
    val: &D::Value,
    assignment: &mut [Option<D::Value>],
    constraints: &[ConstraintEnum<D>],
    adjacency: &Adjacency,
) -> usize
where
    D::Value: PartialEq + Clone,
{
    // Temporarily patch var's slot, count violations, then restore.
    let original = assignment[var as usize].take();
    assignment[var as usize] = Some(val.clone());

    let mut count = 0;
    for &ci in adjacency.constraints_for(var) {
        if !constraints[ci as usize].check(assignment) {
            count += 1;
        }
    }

    // Restore.
    assignment[var as usize] = original;
    count
}

/// Return the set of variable ids that participate in at least one violated
/// constraint under the current (complete) assignment.
fn conflicting_variables<D: Domain>(
    assignment: &[Option<D::Value>],
    constraints: &[ConstraintEnum<D>],
    _adjacency: &Adjacency,
) -> Vec<VarId>
where
    D::Value: PartialEq + Clone,
{
    let num_vars = assignment.len();
    let mut in_conflict = vec![false; num_vars];

    for constraint in constraints {
        if !constraint.check(assignment) {
            for &v in constraint.scope() {
                in_conflict[v as usize] = true;
            }
        }
    }

    (0..num_vars as VarId)
        .filter(|&v| in_conflict[v as usize])
        .collect()
}

// ---------------------------------------------------------------------------
// Min-conflicts solver
// ---------------------------------------------------------------------------

/// Run min-conflicts local search.
///
/// Starts from a random complete assignment drawn from each variable's current
/// domain, then iteratively picks a random conflicting variable and reassigns
/// it to the value that minimises the number of violated constraints.
///
/// Returns `Some(assignment)` if a solution is found within `max_iterations`,
/// or `None` if the budget is exhausted.
pub fn min_conflicts<D: Domain>(
    variables: &mut [Variable<D>],
    constraints: &[ConstraintEnum<D>],
    adjacency: &Adjacency,
    max_iterations: usize,
) -> Option<Vec<D::Value>>
where
    D::Value: PartialEq + Clone,
{
    let n = variables.len();
    if n == 0 {
        return Some(Vec::new());
    }

    let mut rng = Rng::new();

    // 1. Random initial assignment.
    let mut assignment: Vec<Option<D::Value>> = Vec::with_capacity(n);
    for var in variables.iter() {
        let vals = var.domain.values();
        if vals.is_empty() {
            return None; // empty domain → unsatisfiable
        }
        let idx = rng.usize(vals.len());
        assignment.push(Some(vals[idx].clone()));
    }

    // 2. Iterative improvement.
    for _ in 0..max_iterations {
        let conflicts = conflicting_variables(&assignment, constraints, adjacency);
        if conflicts.is_empty() {
            // All constraints satisfied — extract solution.
            return Some(assignment.into_iter().map(|v| v.unwrap()).collect());
        }

        // Pick a random conflicting variable.
        let var = conflicts[rng.usize(conflicts.len())];
        let domain_vals = variables[var as usize].domain.values();
        if domain_vals.is_empty() {
            return None;
        }

        // Find the value that minimises conflicts.  On ties, keep the first
        // encountered (which is effectively random given domain iteration order
        // and the random initial state).
        let mut best_val = domain_vals[0].clone();
        let mut best_count = usize::MAX;

        for val in &domain_vals {
            let c = num_conflicts::<D>(var, val, &mut assignment, constraints, adjacency);
            if c < best_count {
                best_count = c;
                best_val = val.clone();
                if best_count == 0 {
                    break; // can't do better
                }
            }
        }

        assignment[var as usize] = Some(best_val);
    }

    // Budget exhausted.
    None
}
